[inductor] Indexing refactors

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116360
* #116079
* __->__ #116078

Perf differences seems to be noise:
![image](https://github.com/pytorch/pytorch/assets/533820/d7a36574-0388-46e4-bd4d-b274d37cab2b)



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

jansel committed:
[inductor] Indexing refactors

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Indexing refactors"


Perf differences seems to be noise:
![image](https://github.com/pytorch/pytorch/assets/533820/d7a36574-0388-46e4-bd4d-b274d37cab2b)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1673,14 +1721,17 @@ def final_argreduce(buffer, result_var, value, index):
         cond = " & ".join(masks)
 
+        def where_cond(tval, fval):
+            if not cond:
```
This check doesn't seem to have existed before, so I'm wondering how could the previous version of the code have worked with an empty `cond` inlined in the literal `"tl.where(...)"`.

jansel replied:
Adding a maximum size for RBLOCK means the mask elimination pass sometimes results in no mask.  Before RBLOCK always existed.

aakhundov commented on torch/_inductor/sizevars.py:
```diff
@@ -359,6 +359,11 @@ def evaluate_static_shape(self, left: Expr) -> int:
         return [self.evaluate_static_shape(x) for x in left]
 
+    def remove_precomputed_replacements(self, expr: Expr) -> Expr:
+        if any(s.name.startswith("ps") for s in expr.free_symbols):
+            return sympy_subs(expr, self.inv_precomputed_replacements)
+        return expr
```
Previously, the removal was recursive until no `ps*` symbols remain to replace. Now it's not: there is just one replacement pass. I guess, if some of the expressions in the values of the `self.inv_precomputed_replacements` contain some of the symbols in the keys, one replacement pass won't be sufficient here?

jansel replied:
I changed `self.inv_precomputed_replacements` so that we apply replacements to the replacement when we create it.  So the loop should no longer be needed.

aakhundov commented on torch/_inductor/virtualized.py:
```diff
@@ -72,6 +72,7 @@ def __init__(self):
         super().__init__()
         self.removed_buffers = set()
         self.inplaced_to_remove = set()
+        self.index_dtype = "tl.int64"
```
This may result in a perf regression in indexing-compute-heavy kernels? Although, your benchmark run suggests it doesn't.

jansel replied:
I think this code is only hit in unit tests when `V.kernel` is set to `NullKernel`.  In codegen it should point to a real kernel, which has this value set via analysis.

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1478,7 +1522,7 @@ def load(self, name: str, index: sympy.Expr):
                 names = {name}
             last_use = len(names & self.last_usage) > 0
-            evict_last = not last_use and ("rmask" in mask or indirect_indexing)
+            evict_last = not last_use and (has_rindex or indirect_indexing)
```
They should be the same, since rmask is added when `any(var.name.startswith("r") for var in index_vars)`

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1249,6 +1278,8 @@ def index_to_str(self, index: sympy.Expr) -> str:
         Rename_indexing and codegen_indexing keep track of the needed indices and add
         new parameters to the function signature.
         """
+        if isinstance(index, list):
```
It can only be a list in this version.

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -79,13 +79,35 @@
     def _print_floor(self, expr):
         assert len(expr.args) == 1
-        return f"tl.math.floor({self.paren(self._print(expr.args[0]))})"
+        return f"tl.math.floor({self._print(expr.args[0])}).to({V.kernel.index_dtype})"
```
Yes, triton will optimize it away.

shunting314 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -822,10 +851,10 @@ def codegen_header(self, code, no_x_dim=False):
             code.writeline(
                 f"{self.name} = {self.ranges_code()}",
             )
+            code.writeline("roffset = 0")
```
Curious why we add this even though roffset will be setup properly in the reduction loop

jansel replied:
For persistent reductions

shunting314 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1488,7 +1532,11 @@ def load(self, name: str, index: sympy.Expr):
+        if (
+            (has_tmpmask or has_rindex)
+            and V.graph.get_dtype(name) != torch.bool
+            and indexing.has_mask()
```
Does `has_tmpmask or has_rindex` implies that `indexing.has_mask()` is true?

shunting314 replied:
Oh, there is corner case where there is a rindex but the rmask is filtered? Realized when reading the discussion below.

shunting314 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -1025,6 +1028,9 @@ def triton_config_tiled_reduction(size_hints, x, y, r, num_stages=1):
     cfg = {"XBLOCK": x, "YBLOCK": y, "RBLOCK": r}
     num_warps = next_power_of_2(min(max(conditional_product(x, y, r) // 256, 1), 8))
     check_config(cfg, xnumel=size_hints[0], ynumel=size_hints[1])
+    assert (
```
Is this function `triton_config_tiled_reduction` even called somewhere?

jansel replied:
No

jansel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -1148,6 +1155,8 @@ def reduction(
     """args to @triton.heuristics()"""
     inductor_meta = {} if inductor_meta is None else inductor_meta
+    if inductor_meta.get("no_x_dim"):
+        size_hints = [1, *size_hints[1:]]
```
Yes that case happens.

jansel committed:
Update on "[inductor] Indexing refactors"


Perf differences seems to be noise:
![image](https://github.com/pytorch/pytorch/assets/533820/d7a36574-0388-46e4-bd4d-b274d37cab2b)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
